### PR TITLE
Feature/reset on error config

### PR DESCRIPTION
**Pull request by CAMOBAP**
- #20

### CODE REVIEW

**Comment by CAMOBAP** [comment]
@e271828- still have some troubles getting `expired-callback` or `` called on Android.

It works without any issues on iOS.

On Android I observed a strange log statement in WebView: `recaptchacompat disabled`, can it be related?

**Comment by e271828-** [comment]
> still have some troubles getting expired-callback or chalexpired-callback called on Android.
> It works without any issues on iOS.

Maybe this is a race against webview close?

> On Android I observed a strange log statement in WebView: recaptchacompat disabled, can it be related?

No: that would be unrelated. You can set it on iOS to confirm.

**Comment by CAMOBAP** [comment]
> Maybe this is a race against webview close?

There is definitely no race condition, here is stack-trace from iOS

Screen Shot 2022-03-01 at 3 35 41 PM

And there is nothing similar to this on Android, what I can observe from my side is that `challenge.chat.listen` wasn't fired

It's hard to analyze minified JS, @e271828-  is it possible to get some dev (non-minified) version of https://js.hcaptcha.com/1/api.js ?

File: sdk/src/main/java/com/hcaptcha/sdk/HCaptchaConfig.java [hunk 1]
```diff
@@ -100,4 +100,10 @@
      */
     @Builder.Default
     private String customTheme = null;
+
+    /**
+     * Reset hCaptcha on errors
+     */
+    @Builder.Default
+    private Boolean resetOnError = false;
```
Comment (DSergiu) [inline sdk/src/main/java/com/hcaptcha/sdk/HCaptchaConfig.java:108]: The issue states to reset on `timeout`. Shall the naming be updated to `resetOnTimeout`?

File: sdk/src/main/res/values-he/strings.xml [hunk 2]
```diff
@@ -1,3 +0,0 @@
-<resources>
-	<string name="loading_text">מאמת שאתה אנושי</string>
```
Comment (DSergiu) [inline sdk/src/main/res/values-he/strings.xml:2]: Why is this translation removed? Seems unrelated to the Issue.

File: sdk/src/main/java/com/hcaptcha/sdk/HCaptchaConfig.java (hunk 1 above)
Comment (CAMOBAP) [reply sdk/src/main/java/com/hcaptcha/sdk/HCaptchaConfig.java:108]: @DSergiu good proposal, BTW maybe we should make it more like iOS and implement actual `resetOnError` not only for timeouts, @e271828- how do you think?

P.S. If not I'm totally agree to rename it

File: sdk/src/main/res/values-he/strings.xml (hunk 2 above)
Comment (CAMOBAP) [reply sdk/src/main/res/values-he/strings.xml:2]: This was done in separate commit, which fixes link issues

File: sdk/src/main/java/com/hcaptcha/sdk/HCaptchaDialogFragment.java [hunk 3]
```diff
@@ -156,6 +157,20 @@ private void setupWebView(final WebView webView) {
         webView.loadUrl("file:///android_asset/hcaptcha-form.html");
     }
 
+    private OnFailureListener getFailureListener(boolean resetOnError) {
+        if (resetOnError) {
+            return new OnFailureListener() {
+                @Override
+                public void onFailure(@NonNull final HCaptchaException exception) {
+                    if (exception.getHCaptchaError() == HCaptchaError.SESSION_TIMEOUT) {
```
Comment (DSergiu) [inline sdk/src/main/java/com/hcaptcha/sdk/HCaptchaDialogFragment.java:165]: It looks like redundant handler as there is already one on line 207. This if statement can be moved there.

File: sdk/src/main/res/values-he/strings.xml (hunk 2 above)
Comment (DSergiu) [reply sdk/src/main/res/values-he/strings.xml:2]: Why are there linting issues? Maybe due to the dependency updates? Seems unrelated to the Issue, please test the 3 updated languages that they still work.

File: sdk/src/main/java/com/hcaptcha/sdk/HCaptchaDialogFragment.java (hunk 3 above)
Comment (DSergiu) [reply sdk/src/main/java/com/hcaptcha/sdk/HCaptchaDialogFragment.java:165]: I tested with updated existing handler and seems to work just fine.

```java
// line 191
@Override
public void onFailure(@NonNull final HCaptchaException hCaptchaException) {
    final boolean silentRetry = this.resetOnTimeout && hCaptchaException.getHCaptchaError() == HCaptchaError.SESSION_TIMEOUT;
    if (isAdded() && !silentRetry) {
        dismiss();
    }
    handler.post(new Runnable() {
        @Override
        public void run() {
            if (silentRetry) {
                // Re-open the challenge
                webView.loadUrl("javascript:hcaptcha.execute()");
            } else {
                hCaptchaDialogListener.onFailure(hCaptchaException);
            }
        }
    });
}
```

Comment (CAMOBAP) [reply sdk/src/main/java/com/hcaptcha/sdk/HCaptchaDialogFragment.java]: Updated

File: sdk/src/main/res/values-he/strings.xml (hunk 2 above)
Comment (CAMOBAP) [reply sdk/src/main/res/values-he/strings.xml:2]: Testing still needs to be done, I will check after https://github.com/hCaptcha/HCaptcha-ios-sdk/issues/50

Comment (CAMOBAP) [reply sdk/src/main/res/values-he/strings.xml:2]: @DSergiu this warning highlight actual problems. What changes was done:

1. `sdk/src/main/res/values-he/strings.xml` -> `sdk/src/main/res/values-iw/strings.xml` (hebrew
2. `sdk/src/main/res/values-id/strings.xml` -> `sdk/src/main/res/values-in/strings.xml` (indonesian)
3. `sdk/src/main/res/values-yi/strings.xml` -> `sdk/src/main/res/values-ji/strings.xml` (yiddish)

Before this change, Indonesian doesn't work at all (it shows English as fallback) I have tested actually device (Redmi Note 7) and emulator

Unfortunatelly I cannot test Yiddish, no such option on my device

**Comment by CAMOBAP** [comment]
Because of this feature, changes HTML significantly, I have tested:
 - Localization for changed strings
 - Tested add three sizes: invisible, normal, compact
 - Timeout happens only for invisible (the same as iOS)

@e271828- @DSergiu let me know if some other cases should be checked

File: sdk/src/main/java/com/hcaptcha/sdk/HCaptchaConfig.java (hunk 1 above)
Comment (e271828-) [reply sdk/src/main/java/com/hcaptcha/sdk/HCaptchaConfig.java]: As long as it's accurate and consistent between iOS/Android, I don't have a preference.

Comment (CAMOBAP) [reply sdk/src/main/java/com/hcaptcha/sdk/HCaptchaConfig.java]: I think `resetOnError` may be the same as on iOS and we will have consistent behavior on both platforms, @DSergiu how do you think?

File: sdk/src/main/assets/hcaptcha-form.html [hunk 4]
```diff
@@ -104,20 +123,9 @@
...
-            if (rqdata) {
-                hcaptcha.setData(hCaptchaID, {
-                    rqdata: rqdata
-                });
-            }
-
-            if (renderConfig.size === 'invisible') {
-                hcaptcha.execute(hCaptchaID);
-            } else {
-                BridgeObject.onLoaded();
```
Comment (DSergiu) [inline sdk/src/main/assets/hcaptcha-form.html:119]: `BridgeObject.onLoaded()` is never called w/ current changes.

File: sdk/src/main/java/com/hcaptcha/sdk/HCaptchaConfig.java [hunk 5]
```diff
@@ -100,4 +100,10 @@
      */
     @Builder.Default
     private String customTheme = null;
+
+    /**
+     * Reset hCaptcha on errors
```
Comment (DSergiu) [inline sdk/src/main/java/com/hcaptcha/sdk/HCaptchaConfig.java:105]: JSDoc says `on error` while var name says `on timeout`.

File: sdk/src/test/java/com/hcaptcha/sdk/HCaptchaJSInterfaceTest.java [hunk 6]
```diff
@@ -116,6 +118,8 @@ public void subset_config_serialization() throws JsonProcessingException, JSONEx
         expected.put("theme", HCaptchaTheme.DARK.toString());
         expected.put("customTheme", JSONObject.NULL);
         expected.put("host", JSONObject.NULL);
+        expected.put("resetOnTimeout", false);
+        expected.put("resetOnError", false);
```
Comment (DSergiu) [inline sdk/src/test/java/com/hcaptcha/sdk/HCaptchaJSInterfaceTest.java:122]: `resetOnError` is unused and it should probably be removed.

File: sdk/src/main/java/com/hcaptcha/sdk/HCaptchaDialogFragment.java [hunk 7]
```diff
@@ -11,6 +11,7 @@
 import android.os.Bundle;
 import android.os.Handler;
 import android.os.Looper;
+import android.util.Log;
```
Comment (DSergiu) [inline sdk/src/main/java/com/hcaptcha/sdk/HCaptchaDialogFragment.java:14]: `Log` import is unused.

File: sdk/src/main/java/com/hcaptcha/sdk/HCaptchaConfig.java (hunk 1 above)
Comment (DSergiu) [reply sdk/src/main/java/com/hcaptcha/sdk/HCaptchaConfig.java]: The 2 options behave differently. We can have both `resetOnError` and `resetOnTimeout` as separate configurable options.

Comment (CAMOBAP) [reply sdk/src/main/java/com/hcaptcha/sdk/HCaptchaConfig.java]: ok, let's stay with `resetOnTimeout` only to not introduce more differences between SDK

File: sdk/src/test/java/com/hcaptcha/sdk/HCaptchaJSInterfaceTest.java (hunk 6 above)
Comment (CAMOBAP) [reply sdk/src/test/java/com/hcaptcha/sdk/HCaptchaJSInterfaceTest.java]: Removed

File: sdk/src/main/java/com/hcaptcha/sdk/HCaptchaDialogFragment.java (hunk 7 above)
Comment (CAMOBAP) [reply sdk/src/main/java/com/hcaptcha/sdk/HCaptchaDialogFragment.java]: Done

File: sdk/src/main/java/com/hcaptcha/sdk/HCaptchaConfig.java (hunk 5 above)
Comment (CAMOBAP) [reply sdk/src/main/java/com/hcaptcha/sdk/HCaptchaConfig.java]: Updated

File: sdk/src/main/assets/hcaptcha-form.html (hunk 4 above)
Comment (CAMOBAP) [reply sdk/src/main/assets/hcaptcha-form.html:119]: Good catch, fixed and retested, invisible, compact and normal sizes

File: sdk/src/main/assets/hcaptcha-form.html [hunk 8]
```diff
@@ -58,6 +58,27 @@
...
+        }
+
+        if (renderConfig.size === 'invisible') {
+            hcaptcha.execute(hCaptchaID);
+        } else {
+            BridgeObject.onLoaded();
+        }
+    }
+
+    function reset() {
```
Comment (DSergiu) [inline sdk/src/main/assets/hcaptcha-form.html:78]: Function name might be misleading. `resetAndExecute()` should be more appropriate.

Comment (CAMOBAP) [reply sdk/src/main/assets/hcaptcha-form.html]: Fixed

Comment (DSergiu) [reply sdk/src/main/assets/hcaptcha-form.html]: Would be nice to avoid force pushes as it makes it harder to diff on what was actually changed since last review. Upon merge, it is fine to do "squash and ff-merge". @CAMOBAP

Comment (CAMOBAP) [reply sdk/src/main/assets/hcaptcha-form.html]: @DSergiu totally makes sense, sorry I forgot that we follow squash and ff-merge
